### PR TITLE
enable multivalues insert

### PR DESCRIPTION
**Pull request by danfrankj**
### Summary
Currently when pushing a dataframe to a database, lines are inserted one by one. This change enables multivalues inserts. 

TODO
 - [x] release note
 - [ ] address chunksize behavior

### Reference
http://docs.sqlalchemy.org/en/rel_0_9/core/dml.html?highlight=insert%20values#sqlalchemy.sql.expression.Insert.values

### CODE REVIEW

**Comment by TomAugspurger** [comment]
xref https://github.com/pandas-dev/pandas/issues/14315 and https://github.com/pandas-dev/pandas/issues/8953

I think the previous issues were with sqlalchemy dialects that don't support multi-row inserts, so we'll need to test that.

+ docs, release note, and tests.

**Comment by danfrankj** [comment]
For reference, I'm using a SQLAlchemy dialect that supports_multivalues_inserts and inserts still happen line by line. Will add tests though.

**Comment by TomAugspurger** [comment]
Ideally, in the tests we'll be able to introspect the sqlalchemy engine somehow to assert that multi-row inserts are actually happening.

**Review by jreback — CHANGES_REQUESTED** [review]
this needs a test

**Comment by danfrankj** [comment]
@jreback @TomAugspurger added first stab at a test. Let me know what you think!

File: pandas/tests/io/test_sql.py [hunk 1]
```diff
@@ -479,6 +479,25 @@ def _transaction_test(self):
         res2 = self.pandasSQL.read_query('SELECT * FROM test_trans')
         assert len(res2) == 1
 
+    def _test_insert_multivalues(self):
```
Comment (TomAugspurger) [inline pandas/tests/io/test_sql.py]: Add a comment with the Github issue number.

Comment (TomAugspurger) [reply pandas/tests/io/test_sql.py]: And since this is sqlalchemy-specific, could you just define the test directly on `_TestSQLAlchemy`?

**Review by TomAugspurger — COMMENTED** [review]
Test seems good, thanks.

How does this interact with `chunksize`? Can we say that you'll have `len(df) // chunksize` inserts? Is there any way to test that?

Also need a release note.

File: pandas/tests/io/test_sql.py (hunk 1 above)
Comment (danfrankj) [reply pandas/tests/io/test_sql.py]: Added github issues as comments and test moved

File: pandas/io/sql.py [hunk 2]
```diff
@@ -572,8 +572,11 @@ def create(self):
         else:
             self._execute_create()
 
-    def insert_statement(self):
-        return self.table.insert()
+    def insert_statement(self, data, conn):
```
Comment (jreback) [inline pandas/io/sql.py]: can you add a doc-string here

File: pandas/io/sql.py [hunk 3]
```diff
@@ -613,7 +616,7 @@ def insert_data(self):
 
     def _execute_insert(self, conn, keys, data_iter):
         data = [{k: v for k, v in zip(keys, row)} for row in data_iter]
-        conn.execute(self.insert_statement(), data)
+        conn.execute(*self.insert_statement(data, conn))
```
Comment (jreback) [inline pandas/io/sql.py]: here as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_sql.py [hunk 4]
```diff
@@ -1665,6 +1665,29 @@ class Temporary(Base):
 
         tm.assert_frame_equal(df, expected)
 
+    def test_insert_multivalues(self):
```
Comment (jreback) [inline pandas/tests/io/test_sql.py]: can you explicity test which backends support this.

**Review by jreback — CHANGES_REQUESTED** [review]
need a whatsnew note that lists the backends where this would work.

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -323,6 +323,7 @@ Other Enhancements
 
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
+- :func:`pd.io.sql.to_sql` now performs a multivalue insert if the underlying connection supports this rather than inserting row by row (:issue:`14315`, :issue: `8953`)
```
Comment (danfrankj) [inline doc/source/whatsnew/v0.23.0.txt]: not sure if this should be an enhancement or a bugfix?

File: pandas/tests/io/test_sql.py (hunk 4 above)
Comment (danfrankj) [reply pandas/tests/io/test_sql.py]: @jreback  I believe I've done this by adding a class variable to the below classes. Let me know if that addresses this concern

**Comment by danfrankj** [comment]
@jreback @TomAugspurger I believe I've addressed your concerns above, when you get a chance PTAL

File: pandas/io/sql.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/io/sql.py]: Could you add a parameters and returns section as well. http://numpydoc.readthedocs.io/en/latest/format.html

File: pandas/io/sql.py [hunk 6]
```diff
@@ -572,8 +572,15 @@ def create(self):
...
-        return self.table.insert()
+    def insert_statement(self, data, conn):
+        """
+        Generate tuple of SQLAlchemy insert statement and any arguments
+        to be executed by connection (via `_execute_insert`).
+        """
+        dialect = getattr(conn, 'dialect', None)
+        if dialect and getattr(dialect, 'supports_multivalues_insert', False):
+            return (self.table.insert(data),)
+        return (self.table.insert(), data)
```
Comment (TomAugspurger) [inline pandas/io/sql.py]: Shouldn't need the parenthesis here.

File: doc/source/whatsnew/v0.23.0.txt [hunk 7]
```diff
@@ -296,6 +296,8 @@ Other Enhancements
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
 - :func:`DataFrame.from_dict` now accepts a ``columns`` argument that can be used to specify the column names when ``orient='index'`` is used (:issue:`18529`)
+- :func:`pd.io.sql.to_sql` now performs a multivalue insert if the underlying connection supports this rather than inserting row by row.
+  SQL dialects supporting multivalue inserts include mysql, postgresql, sqlite and any dialect with `supports_multivalues_insert`. (:issue:`14315`, :issue:`8953`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: "dialect" -> "SQLAlchemy dialect"

File: doc/source/whatsnew/v0.23.0.txt [hunk 8]
```diff
@@ -296,6 +296,8 @@ Other Enhancements
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
 - :func:`DataFrame.from_dict` now accepts a ``columns`` argument that can be used to specify the column names when ``orient='index'`` is used (:issue:`18529`)
+- :func:`pd.io.sql.to_sql` now performs a multivalue insert if the underlying connection supports this rather than inserting row by row.
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: I don't know if `io.sql.to_sql` is part of the API list. Better to just have

    :meth:`DataFrame.to_sql`

**Review by TomAugspurger — COMMENTED** [review]
Just some minor comments. Looks good overall.

Can you run a quick benchmark to see how things look compared to master? How much faster are things?

**Comment by danfrankj** [comment]
Some profiling 

### Presto

```

In [5]: df = pd.DataFrame(np.random.randn(1000, 4), columns=list('ABCD'))

In [6]: %time df.to_sql('multi_insert_profile', presto_engine, schema='dan_frank', index=False)
INFO:pyhive.presto:SHOW COLUMNS FROM "dan_frank"."multi_insert_profile"
INFO:pyhive.presto:
CREATE TABLE "dan_frank"."multi_insert_profile" (
    "A" DOUBLE,
    "B" DOUBLE,
    "C" DOUBLE,
    "D" DOUBLE
)


INFO:pyhive.presto:INSERT INTO "dan_frank"."multi_insert_profile" ("A", "B", "C", "D") VALUES (-0.4425108530531213, -0.4582021047086419, 0.17242001384630398, -1.2917653645626361), (-0.9715964127007015, -0.1458055798883143, 0.3444250700373072, -0.35869901840257923), (0.6732070093449385, 0.3371601918897362, -0.49645678476330574, -0.8241023338536242), (-0.4845513289740901, -1.4860936235542728, 0.19123940403655423, -0.32166319533058985), (2.72221337305179, 0.31572155167450705, -0.5522159042533455, -0.28023622560479866), (-2.2406710854261345, 0.8005522925313067, -0.5762370339886204, 1.1784968768877826), (-0.06826129801094293, 0.2760723638718846, 0.526970720133034, 
... LOG TRUNCATED


CPU times: user 131 ms, sys: 9.06 ms, total: 140 ms
Wall time: 13.1 s



In [7]: presto_engine.dialect.supports_multivalues_insert = False

In [8]: %time df.to_sql('sequential_insert_profile', presto_engine, schema='dan_frank', index=False)
INFO:pyhive.presto:SHOW COLUMNS FROM "dan_frank"."sequential_insert_profile"
INFO:pyhive.presto:
CREATE TABLE "dan_frank"."sequential_insert_profile" (
    "A" DOUBLE,
    "B" DOUBLE,
    "C" DOUBLE,
    "D" DOUBLE
)


INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (-0.4425108530531213, -0.4582021047086419, 0.17242001384630398, -1.2917653645626361)
INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (-0.9715964127007015, -0.1458055798883143, 0.3444250700373072, -0.35869901840257923)
INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (0.6732070093449385, 0.3371601918897362, -0.49645678476330574, -0.8241023338536242)
INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (-0.4845513289740901, -1.4860936235542728, 0.19123940403655423, -0.32166319533058985)
INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (2.72221337305179, 0.31572155167450705, -0.5522159042533455, -0.28023622560479866)
INFO:pyhive.presto:INSERT INTO "dan_frank"."sequential_insert_profile" ("A", "B", "C", "D") VALUES (-2.2406710854261345, 0.8005522925313067, -0.5762370339886204, 1.1784968768877826)
... LOG TRUNCATED 


CPU times: user 15.7 s, sys: 1.44 s, total: 17.1 s
Wall time: 14min 57s
























































```

**Comment by danfrankj** [comment]
@TomAugspurger did some brief profiling and added additional documentation. Anything else you think is needed for this PR?

**Comment by TomAugspurger** [comment]
@jorisvandenbossche any thoughts? This seems harmless to me, and the results are... impressive :)

File: doc/source/whatsnew/v0.23.0.txt [hunk 9]
```diff
@@ -340,6 +340,8 @@ Other Enhancements
 - Added option ``display.html.use_mathjax`` so `MathJax <https://www.mathjax.org/>`_ can be disabled when rendering tables in ``Jupyter`` notebooks (:issue:`19856`, :issue:`19824`)
 - :meth:`Timestamp.month_name`, :meth:`DatetimeIndex.month_name`, and :meth:`Series.dt.month_name` are now available (:issue:`12805`)
 - :meth:`Timestamp.day_name` and :meth:`DatetimeIndex.day_name` are now available to return day names with a specified locale (:issue:`12806`)
+- :meth:`DataFrame.to_sql` now performs a multivalue insert if the underlying connection supports this rather than inserting row by row.
+  SQLAlchemy dialects supporting multivalue inserts include mysql, postgresql, sqlite and any dialect with `supports_multivalues_insert`. (:issue:`14315`, :issue:`8953`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: double back-ticks. can you also add a 'note' in io.rst about this.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by danfrankj** [comment]
@jreback backquotes fixed and note added to `io.rst`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by danfrankj** [comment]
Thank you guys for the reviews! Excited for my first pandas contribution :)

**Comment by TomAugspurger** [comment]
@tripkane maybe make a new issue (link back here) with a reproducible example.

**Comment by tripkane** [comment]
@TomAugspurger: ok thanks, will do

**Pull request merged**